apis/topology/v1beta1/topology_types.go

```diff
@@ -0,0 +1,108 @@
+/*
+Copyright 2024.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package v1beta1
+
+import (
+	"context"
+	"github.com/openstack-k8s-operators/lib-common/modules/common/helper"
+	"github.com/openstack-k8s-operators/lib-common/modules/common/util"
+	"k8s.io/apimachinery/pkg/util/validation/field"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+)
+
+// TopologySpec defines the desired state of Topology
+type TopologySpec struct {
+	// +kubebuilder:validation:Optional
+	// TopologySpreadConstraints exposes topologySpreadConstraints that are
+	// applied to the StatefulSet
+	TopologySpreadConstraints *[]corev1.TopologySpreadConstraint `json:"topologySpreadConstraints,omitempty"`
+
+	// Affinity exposes PodAffinity, PodAntiaffinity and NodeAffinity overrides
+	// that are applied to StatefulSet/Deployments
+	// +optional
+	Affinity *corev1.Affinity `json:"affinity,omitempty"`
+
+	//TODO: We could add NodeSelector here as it belongs to the same APIGroup
+}
+
+// TopologyStatus defines the observed state of Topology
+type TopologyStatus struct {
+
+	// ObservedGeneration - the most recent generation observed for this
+	// service.
+	ObservedGeneration int64 `json:"observedGeneration,omitempty"`
```

Comment:
As there is no controller reconciling the CR this field will never be updated and therefore will not be in sync with the the generation in the metadata. For an outside observer this signals that the world is not updated according the latest changes in the spec. I suggest to remove this field to avoid that confusion. (or add a controller that does the reconciliation to update this field when the spec changes)

Reply:
are you suggesting to remove the `.Status` entirely as there's no data associated with it? Considering it would be an empty struct (and never populated), I would remove it entirely.

Reply:
Removed.

Reply:
My original suggestion was to remove the observerdgeneration field. But if removing the whole status does not cause any trouble then sure I'm OK to drop the whole status too.